ambry-api/src/main/java/com.github.ambry/network/NetworkSendMetrics.java

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.network;
+
+import com.codahale.metrics.Histogram;
+
+
+/**
+ * Tracks a set of metrics for a {@link NetworkSend}
+ */
+public class NetworkSendMetrics {
+  private long timeSpentTillNow;
+  private Histogram requestOrResponseQueueTime;
+  private Histogram requestOrResponseSendTime;
+  private Histogram requestOrResponseTotalTime;
+  private Histogram requestOrResponseSendTimeBySize;
+  private Histogram requestOrResponseTotalTimeBySize;
+
+  public NetworkSendMetrics(Histogram requestOrResponseQueueTime, Histogram requestOrResponseSendTime,
+      Histogram requestOrResponseTotalTime, Histogram requestOrResponseSendTimeBySize,
+      Histogram requestOrResponseTotalTimeBySize, long timeSpentTillNow) {
+    this.requestOrResponseQueueTime = requestOrResponseQueueTime;
+    this.requestOrResponseSendTime = requestOrResponseSendTime;
+    this.requestOrResponseTotalTime = requestOrResponseTotalTime;
+    this.timeSpentTillNow = timeSpentTillNow;
+    this.requestOrResponseSendTimeBySize = requestOrResponseSendTimeBySize;
+    this.requestOrResponseTotalTimeBySize = requestOrResponseTotalTimeBySize;
+  }
+
+  /**
+   * Updates the time spent by the request or response in the queue before being sent out
+   * @param value the time spent by the request or response in the queue
+   */
+  public void updateQueueTime(long value) {
+    requestOrResponseQueueTime.update(value);
+    timeSpentTillNow += value;
+  }
+
+  /**
+   * Updates the time spent by request or response to be completely sent
+   * @param value the time spent by the request or response to be completely sent
+   */
+  public void updateSendTime(long value) {
+    requestOrResponseSendTime.update(value);
+    if (requestOrResponseSendTimeBySize != null) {
+      requestOrResponseSendTimeBySize.update(value);
```

Comment:
Yeah, which doesn't really make sense for this class though. If you were to describe these variables in this class, how would you do it?
I kinda see how something like `MetricsHistogram` would be useful here. Basically, the caller wants to update two different metrics and that can be encapsulated within `MetricsHistogram` class.
I think what you should do is to keep the things that make sense for `NetworkSendMetrics` within this class. Let `NetworkClient` use this class directly as I don't think it needs any other metric. As far as the server is concerned, it can extend this class to make a server specific `NetworkSendMetrics` class which can override appropriate methods.